File: eco-core/core-plugin/src/main/java/com/willfp/reforges/reforges/util/ReforgeArgParser.java
package com.willfp.reforges.reforges.util;

import com.willfp.eco.core.items.args.LookupArgParser;
import com.willfp.reforges.reforges.Reforge;
import com.willfp.reforges.reforges.Reforges;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.jetbrains.annotations.NotNull;

import java.util.function.Predicate;

public class ReforgeArgParser implements LookupArgParser {
    @Override
    public Predicate<ItemStack> parseArguments(@NotNull final String[] args,
                                               @NotNull final ItemMeta meta) {
        Reforge reforge = null;
        for (String arg : args) {
            String[] split = arg.split(":");
            if (split.length == 1 || !split[0].equalsIgnoreCase("reforge")) {
                continue;
            }

            Reforge match = Reforges.getByKey(split[1].toLowerCase());
            if (match == null) {
                continue;
            }

            reforge = match;
            break;
        }

        if (reforge == null) {
            return itemStack -> true;
        }

        ReforgeUtils.setReforge(meta, reforge);

        Reforge finalReforge = reforge;
        return test -> {
            ItemMeta testMeta = test.getItemMeta();
            if (testMeta == null) {
                return false;
            }

            return finalReforge.equals(ReforgeUtils.getReforge(testMeta));
        };
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/reforges/reforges/util/ReforgeArgParser.kt
package com.willfp.reforges.reforges.util

import com.willfp.eco.core.items.args.LookupArgParser
import com.willfp.reforges.reforges.Reforge
import com.willfp.reforges.reforges.Reforges
import com.willfp.reforges.util.reforge
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import java.util.function.Predicate

object ReforgeArgParser : LookupArgParser {
    override fun parseArguments(
        args: Array<String>,
        meta: ItemMeta
    ): Predicate<ItemStack>? {
        var reforge: Reforge? = null

        for (arg in args) {
            val split = arg.split(":").toTypedArray()
            if (split.size == 1 || !split[0].equals("reforge", ignoreCase = true)) {
                continue
            }
            val match = Reforges.getByKey(split[1].lowercase()) ?: continue
            reforge = match
            break
        }

        reforge ?: return null

        meta.reforge = reforge

        return Predicate { test ->
            val testMeta = test.itemMeta ?: return@Predicate false
            reforge == testMeta.reforge
        }
    }

    override fun serializeBack(meta: ItemMeta): String? {
        val reforge = meta.reforge ?: return null

        return "reforge:${reforge.id}"
    }
}
